libsrc/c128/tgi/c128-hi.s

```diff
@@ -0,0 +1,880 @@
+        sta     POINT+1
+        rts
+
+@DECPTR:                     ;Okay, subtract 320 then
+        ldy     #7           ;Y decreased through 0
+        lda     POINT
+        sec
+        sbc     #<320
+        sta     POINT
+        lda     POINT+1
+        sbc     #>320
+        sta     POINT+1
+        rts
+
+; ------------------------------------------------------------------------
+; BAR: Draw a filled rectangle with the corners X1/Y1, X2/Y2, where
+; X1/Y1 = ptr1/ptr2 and X2/Y2 = ptr3/ptr4 using the current drawing color.
+; Contrary to most other functions, the graphics kernel will sort and clip
+; the coordinates before calling the driver, so on entry the following
+; conditions are valid:
+;       X1 <= X2
+;       Y1 <= Y2
+;       (X1 >= 0) && (X1 < XRES)
+;       (X2 >= 0) && (X2 < XRES)
+;       (Y1 >= 0) && (Y1 < YRES)
+;       (Y2 >= 0) && (Y2 < YRES)
+;
+; Must set an error code: NO
+;
+
+; Note: This function needs optimization. It's just a cheap translation of
+; the original C wrapper and could be written much smaller (besides that,
+; calling LINE is not a good idea either).
+
+BAR:
+        lda     X2
+        sta     X2SAVE
+        lda     X2+1
+        sta     X2SAVE+1
+
+        lda     Y2
+        sta     Y2SAVE
+
+        lda     X1
+        sta     X1SAVE
+        lda     X1+1
+        sta     X1SAVE+1
+
+        lda     Y1
+        sta     Y1SAVE
+
+@L1:    sta     Y2
+        lda     #>200
+        sta     Y1+1
+        sta     Y2+1
+
+        jsr     LINE
+
+        lda     Y1SAVE
+        cmp     Y2SAVE
+        beq     @L4
+
+        inc     Y1SAVE
+
+        lda     X1SAVE
+        sta     X1
+        lda     X1SAVE+1
+        sta     X1+1
+
+        lda     X2SAVE
+        sta     X2
+        lda     X2SAVE+1
+        sta     X2+1
+
+        lda     Y1SAVE
+        sta     Y1
+        jmp     @L1
+
+@L4:    rts
+
+
+; ------------------------------------------------------------------------
+; TEXTSTYLE: Set the style used when calling OUTTEXT. Text scaling in X and Y
+; direction is passend in X/Y, the text direction is passed in A.
+;
+; Must set an error code: NO
+;
+
+TEXTSTYLE:
+        stx     TEXTMAGX
+        sty     TEXTMAGY
+        sta     TEXTDIR
+        rts
+
+
+; ------------------------------------------------------------------------
+; OUTTEXT: Output text at X/Y = ptr1/ptr2 using the current color and the
+; current text style. The text to output is given as a zero terminated
+; string with address in ptr3.
+;
+; Must set an error code: NO
+;
+
+OUTTEXT:
+
+; Calculate a pointer to the representation of the character in the
+; character ROM
+
+        ldx     #((>(CHARROM + $0800)) >> 3)
+        ldy     #0
+        lda     (TEXT),y
+        bmi     @L1
+        ldx     #((>(CHARROM + $0000)) >> 3)
+@L1:    stx     ptr4+1
+        asl     a
+        rol     ptr4+1
+        asl     a
+        rol     ptr4+1
+        asl     a
+        rol     ptr4+1
+        sta     ptr4
+
```

Comment:
I don't think that these blank lines belong here.

Reply:
This part is everywhere incomplete, we should insert a TODO.

Reply:
I see.